### ISSUE TITLE
jarsigner -  The signer's certificate is self-signed. No -tsa or -tsacert is provided and this jar is not timestamped.

### ISSUE THREAD

**Issue by gmck** [xamarin/xamarin-android#3637]
Earlier this morning I uploaded an aab created with VS 2019 Professional 16.3.0 Preview 3.0 to the Play store without problems. 

I then uninstalled because the license was due to expire tomorrow and installed VS2019 Community Preview 16.3.0. Preview 3.0.

I now get the above error with the jarsigner when using the Community version. The certificate is my original certificate from 4 years ago.

I uploaded this aab to the Play store, it was accepted and it has now gone live. 

The following is the sequence from the build log re the signing etc.

Line # 160933
BundleConfig.json: {"compression":{"uncompressedGlob":["typemap.mj","typemap.jm","assemblies/*"]}}
C:\Program Files\Android\Jdk\microsoft_dist_openjdk_1.8.0.25\bin\java.exe -Xmx1G -jar "C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\MSBuild\Xamarin\Android\bundletool.jar" 
build-bundle 
--modules obj\Release\90\android\bin\base.zip 
--output obj\Release\90\android\bin\com.glmsoftware.OBDNow.aab 
--config C:\Users\Graham\AppData\Local\Temp\tmpCB0A.tmp

Line # 161025
C:\Program Files\Android\Jdk\microsoft_dist_openjdk_1.8.0.25\bin\jarsigner.exe -keystore F:\glmkeystore\glm.keystore -storepass <password> -keypass <password> -digestalg SHA-256 -sigalg SHA256withRSA -signedjar 
bin\Release\com.glmsoftware.OBDNow-Signed.aab D:\ProjectsVS2019Preview\OBDIINow\OBDIINow\obj\Release\90\android\bin\com.glmsoftware.OBDNow.aab <aliasname> 
jar signed.
The signer's certificate is self-signed.
No -tsa or -tsacert is provided and this jar is not timestamped. Without a timestamp, users may not be able to validate this jar after the signer certificate's expiration date (2034-06-12) or after any future revocation date.

Line # 162619
C:\Program Files\Android\Jdk\microsoft_dist_openjdk_1.8.0.25\bin\java.exe -Xmx1G -jar "C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\MSBuild\Xamarin\Android\bundletool.jar" 
build-apks 
--connected-device 
--bundle obj\Release\90\android\bin\com.glmsoftware.OBDNow.aab 
--output obj\Release\90\android\bin\com.glmsoftware.OBDNow.apks 
--mode default --adb "C:\Program Files (x86)\Android\android-sdk\platform-tools\adb.exe" 
--aapt2 "C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\MSBuild\Xamarin\Android\aapt2.exe" 
--ks F:\glmkeystore\glm.keystore 
--ks-key-alias <aliasname> 
--key-pass pass:<password> 
--ks-pass pass:<password>

Line # 162628
C:\Program Files\Android\Jdk\microsoft_dist_openjdk_1.8.0.25\bin\java.exe -Xmx1G -jar "C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\MSBuild\Xamarin\Android\bundletool.jar" 
install-apks 
--apks obj\Release\90\android\bin\com.glmsoftware.OBDNow.apks 
--adb "C:\Program Files (x86)\Android\android-sdk\platform-tools\adb.exe" 
--allow-downgrade 
--modules _ALL_

**Comment by dellis1972** [comment]
The `No -tsa or -tsacert is provided and this jar is not timestamped.` is normally just a warning. Do you have `Treat Warnings and Errors` enabled on your project? 

If not can you get us the full diagnostic build output for your project. You can do this in Visual Studio using the following.

1. Set **Tools &gt; Options &gt; Project and Solutions \[section\] &gt; Build and Run \[subsection\] &gt; MSBuild project build output verbosity** to **Diagnostic**.
2. Clean your app, build and deploy it to device, and then follow any steps needed in the app to reproduce the error.
3. Open the **View &gt; Output** window, and set **Show output from** to **Build**.
4. Make sure the **Output** window is selected, and then choose **File &gt; Save Output As** to save the contents to a text file. Or copy and paste the build output into a text file.
5. Go back to the **View &gt; Output** window, and set **Show output from** to **Debug**. 
6. Once again make sure the **Output** window is selected, and then choose **File &gt; Save Output As** to save the contents to a text file (using a different name from step 4). Or copy and paste the build output into a text file.
7. Attach back the 2 text files in a new comment on this item, being sure to set **Viewable by** to **Microsoft Only** on your comment before submitting it.

**Comment by gmck** [comment]
@dellis1972  The project has warnings as Errors: none.
I have the build log ready for you but don't see a way here of setting the log file as Viewable by Microsoft only. Could you please advise how?

**Comment by dellis1972** [comment]
You can try emailing it to me. This is a public issue so there are no permissions on attachments as far as I know.

**Comment by dellis1972** [comment]
I just remembered we have seen this before. 

The solution then was to provide the `JarsignerTimestampAuthorityUrl` or `JarsignerTimestampAuthorityCertificateAlias` MSBuild properties which we should then use with JarSigner as outlined in the Java docs https://docs.oracle.com/javase/7/docs/technotes/guides/security/time-of-signing.html.

e.g

```
<JarsignerTimestampAuthorityUrl >http://timestamp.digicert.com</JarsignerTimestampAuthorityUrl >
```

or 

```
/p: JarsignerTimestampAuthorityUrl=http://timestamp.digicert.com
```

**Comment by brendanzagaeski** [comment]
*Cross-referencing note*

<https://github.com/xamarin/xamarin-android/pull/3685> is closely related. That pull request changes the handling of these `jarsigner` messages to surface them as informational messages instead of warnings. That change is included in Visual Studio 2019 version 16.4 Preview 3 and higher.
